src/pyvesync/vesync.py

```diff
@@ -352,33 +352,31 @@ def login(self) -> bool:
 
         Returns:
             True if login successful, False if not
+
+        Raises:
+            VesyncLoginError: If login fails
         """
-        user_check = isinstance(self.username, str) and len(self.username) > 0
-        pass_check = isinstance(self.password, str) and len(self.password) > 0
-        if user_check is False:
-            logger.error('Username invalid')
-            return False
-        if pass_check is False:
-            logger.error('Password invalid')
-            return False
+        if not isinstance(self.username, str) or len(self.username) == 0 \
+                or not isinstance(self.password, str) or len(self.password) == 0:
+            raise VesyncLoginError('Invalid username or password')
 
         response, _ = Helpers.call_api(
             '/cloud/v1/user/login', 'post',
             json_object=Helpers.req_body(self, 'login')
         )
 
-        if Helpers.code_check(response) and 'result' in response:
-            self.token = response.get('result').get('token')
-            self.account_id = response.get('result').get('accountID')
-            self.country_code = response.get('result').get('countryCode')
+        if isinstance(response, dict) and response.get('code') == 0:
+            self.token = response.get('result', {}).get('token')
+            self.account_id = response.get('result', {}).get('accountID')
+            self.country_code = response.get('result', {}).get('countryCode')
             self.enabled = True
             logger.debug('Login successful')
-            logger.debug('token %s', self.token)
-            logger.debug('account_id %s', self.account_id)
-
             return True
-        logger.error('Error logging in with username and password')
-        return False
+        if isinstance(response, dict) and isinstance(response.get('msg'), str):
+            message = response['msg']
+        else:
+            message = 'Unknown error logging in'
+        raise VesyncLoginError(message)
```

Comment:
Down the road it would be great to detect wrong username / pw with a different exception so we can trigger the correct reauth flow in HA vs a API outage.

Reply:
helpers.py line 376 - rate limit error is determined in the call_api method so it applies to every class that calls it. This is just preliminary implementation, it will change with async